### PR TITLE
Replace TTS with Client-Side Espeak

### PR DESCRIPTION
**Pull request by Elijahrane**
PR Purpose: Replaces the old server-side MaryTTS with the compact Espeak system. This is just the initial implementation and soon there will be new voices and modifications of them based on player-controlled in-game factors.

Addresses Issue #1525

What was tested and works:
Linux Client
Windows Client
2 Player Multiplayer with Late Joiners
Input Sanitation

What was not tested:
OSX. 

Additional work:
I couldn't find anyone to compile the OSX espeak system. It should work though, as I've put the code in place in the Unity side already for OSX. Just add the OSX is a folder mirroring the Linux and Windows ones in Assets/StreamingAssets

### CODE REVIEW

File: UnityProject/Assets/StreamingAssets/Espeak/Linux/include/espeak-ng/encoding.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see: <http://www.gnu.org/licenses/>.
+ */
+#ifndef ESPEAK_NG_ENCODING_H
```
Comment (fomalsd) [inline UnityProject/Assets/StreamingAssets/Espeak/Linux/include/espeak-ng/encoding.h]: is it necessary to include all these C files?

Comment (Elijahrane) [reply UnityProject/Assets/StreamingAssets/Espeak/Linux/include/espeak-ng/encoding.h]: Probably not. I haven't pruned the install folder too much

**Comment by daalta** [comment]
Looks cool, a client-side TTS would really be something.

It looks like you've included a lot of TTS languages, too! Is that intentional? Or is it just easier to include them all?

**Comment by Elijahrane** [comment]
They're a few MB and useful for a few things (accent filters, localization). If you want to remove everything but the English I wouldn't be opposed

**Comment by Bod9001** [comment]
just  saying it doesn't work in editor
image

Got that in build
image


All windows

**Comment by Elijahrane** [comment]
Packaging issue. If you find the library it will work. I didn't make the windows build of espeak nor do I know if we're allowed to distribute those libs

**Comment by Elijahrane** [comment]
https://github.com/espeak-ng/espeak-ng/blob/master/docs/building.md

It seems that that redistributable is a dependency for Windows. I guess it's a matter of including it.

https://www.microsoft.com/en-us/download/details.aspx?id=48145

**Comment by Elijahrane** [comment]
What's left:
1.Have someone compile and hook up the Mac Folder.
2.See if there is any advantage in using speak over espeak, or figure out if we could use the shared library folder.
3. Instead of the program running via it's own audio output system, have it give stdout into the old MarryTTS synthesize subroutine for more consistent audio output. Depending on how the mac compiler deals with PCAudioLib this may or may not be required for it to work on mac. Also we can further modulate it that way which we can't do otherwise.

As it is right now it's fully functional though. Macs just go through the old MaryTTS temporarily.

**Comment by Bod9001** [comment]
image
I'm getting this in editor

**Comment by Elijahrane** [comment]
When it checks Application_platform it looks for the player client, not editor.

**Comment by Elijahrane** [comment]
It should work now. At least, every case should be given a file name.

To debug, look at any Win32 exception "File not found". Compare the place it looked and change the filename under startEspeak to correct it.

**Comment by Bod9001** [comment]
humm One last note can you push the metafiles that it changes since unity uses that to track the folders and it regenerates it 
every time I open it

**Comment by Bod9001** [comment]
image
More errors?

**Comment by Elijahrane** [comment]
I thought the Windows binary might have relative pathing because it worked with no arguments when I tested it in Wine. Strange it'd look in a unix directory by default. It needs the --path argument passed like Linux and Mac. I'll fix it after work.

**Comment by Elijahrane** [comment]
The issue is that `espeak.StartInfo.Arguments = "--path=" + Application.dataPath + relative_Path + message;` generates arguments Linux parses correctly, but Windows parses as one giant string to attempt to read and speak. If anyone knows why ProcessStartInfo.Arguments behaves differently on Linux and Windows and how to fix it, I'd appreciate your help.

**Comment by Bod9001** [comment]
humm Maybe try passing \\" and experiment with that a bit?

**Comment by Elijahrane** [comment]
Can you uncomment the Debug.Log I commented out, update scripts, then run and copy + paste the "This was passed to espeak: " in the log and see what it passed?

**Comment by Bod9001** [comment]
ok will do

**Comment by Bod9001** [comment]
~~humm Do this in what script?~~ Forgot to save it xD

**Comment by Bod9001** [comment]
This was passed to espeak: --path=Q:/Fast programmes/Unity station_unity station/unitystation/UnityProject/Assets/StreamingAssets/Espeak/Windows/  "yo this work"

You might need quotation marks around path Since there are spaces in it

**Comment by hecksadecimal** [comment]
Oh it's just one directory away from being correct, I think.
Edit: Disregard. It's probably just that the path needs to be wrapped in quotes.

**Comment by Elijahrane** [comment]
Try now on Windows. Also for Mac testers notice that the legacy TTS Is only flagged for the mac player and not editor.

**Comment by hecksadecimal** [comment]
So two things. 
One, it's still giving that same error. 
Two, when it tries to launch espeak, it opens up a windows console that very briefly gets in the way of Unitystation.

**Comment by Elijahrane** [comment]
I managed to get the threading working properly and implemented Windows parsing and unintended windows popping up with help from hecksadecimal. In addition to fixing the Windows hang (I hope, at least) threading should also lessen performance impact.

**Comment by Bod9001** [comment]
ok cool, Seems to be working I can hear it, only issue now is to test it to make sure you can't put characters into run stuff on people's computers

**Comment by Elijahrane** [comment]
Screenshot from 2020-02-13 07-56-51
😎

**Comment by fomalsd** [comment]
Tried a mac github build and TTS used is still the serverside MaryTTS. Maybe it didn't build properly for some reason...

**Comment by hecksadecimal** [comment]
I think the mac stuff might've been commented out until a mac build of the TTS was made available.

**Comment by fomalsd** [comment]
I really think TTS should be consistent on all platforms. Have you tried espeak-ng?

**Comment by Elijahrane** [comment]
I asked around for someone to compile for mac and no one took up the offer. I could cross compile but there'd be no one to test.

**Comment by Elijahrane** [comment]
To be clear, the Linux and Windows binaries were manually compiled.

**Comment by fomalsd** [comment]
While I really appreciate the effort, espeak simply doesn't sound good enough (you can listen to examples in discord), and slapping reverb on it to try to make it sound less harsh doesn't sound like a proper solution. I'm going to close this one.

**Comment by daalta** [comment]
This is a good place to test out what espeak sounds like.

https://eeejay.github.io/espeak/emscripten/espeak.html

I don't use TTS, so it's hard for me to judge how distracting or useful espeak is compared to our current system. MaryTTS definitely sounds more natural but eSpeak runs client side. I feel like neither solution is perfect. Should we really rule out eSpeak?

**Comment by fomalsd** [comment]
Rane provided examples of something else what sounded quite alright, I think it was called Festival

**Comment by daalta** [comment]
Yes, I've stumbled upon festival too. It sounds more natural than espeak.

It looked like it's not being developed as actively though. At least from a first glance.

**Pull request closed without merging**